### PR TITLE
SPARKC-161: RDD coalesce() pushdown to cassandra.

### PR DESCRIPTION
**Pull request by artem-aliev**
The coalesce method looks more convinient to specify the number of paritions than ReadConf usage.

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala [hunk 1]
```diff
@@ -77,6 +78,21 @@ class CassandraTableScanRDD[R] private[connector](
...
+  private def copy(numPartitions: Option[Int]): Self = {
+    new CassandraTableScanRDD[R](
+      sc = sc,
+      connector = connector,
+      keyspaceName = keyspaceName,
+      tableName = tableName,
+      columnNames = columnNames,
+      where = where,
+      limit = limit,
+      numPartitions = numPartitions,
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: Could we just manipulate the read conf's splitCount rather than introducing a new variable here?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala [hunk 2]
```diff
@@ -221,10 +238,36 @@ class CassandraTableScanRDD[R] private[connector](
     if (containsPartitionKey(where)) {
       CassandraPartitionGenerator(connector, tableDef, Some(1), splitSize)
     } else {
-      CassandraPartitionGenerator(connector, tableDef, splitCount, splitSize)
+      CassandraPartitionGenerator(connector, tableDef, numPartitions.orElse(splitCount), splitSize)
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: This logic would be simplified as well if we modified the read conf instead of adding a new "numPartitions"

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala [hunk 3]
```diff
@@ -221,10 +238,36 @@ class CassandraTableScanRDD[R] private[connector](
...
     }
   }
 
+  /**
+    * The method does not create CoalesceRDD but reduce number of parittions to read from Cassandra.
+    * It turn off partition size calcutlation and ignore spark.cassandra.input.split.size
+    * The method is useful with where() method call, when actual size of data is smaller then the table size.
+    * It has no effect if a partition key is used in where clause.
+    *
+    * @param numPartitions number of partinions
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: sp on partinions

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala [hunk 4]
```diff
@@ -221,10 +238,36 @@ class CassandraTableScanRDD[R] private[connector](
...
+    * @param shuffle       weither to call shufle after
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: sp on weither -> whether

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala [hunk 5]
```diff
@@ -221,10 +238,36 @@ class CassandraTableScanRDD[R] private[connector](
...
+    * @param ord
+    * @return new CassandraTableScanRDD with predefined nube of paritions
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: sp on nube -> number, parit -> partitions

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala (hunk 4 above)
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: and shuffle

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala [hunk 6]
```diff
@@ -221,10 +238,36 @@ class CassandraTableScanRDD[R] private[connector](
     if (containsPartitionKey(where)) {
       CassandraPartitionGenerator(connector, tableDef, Some(1), splitSize)
     } else {
-      CassandraPartitionGenerator(connector, tableDef, splitCount, splitSize)
+      CassandraPartitionGenerator(connector, tableDef, numPartitions.orElse(splitCount), splitSize)
     }
   }
 
+  /**
+    * The method does not create CoalesceRDD but reduce number of parittions to read from Cassandra.
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: Maybe reword

"This method overrides the default spark behavior and will not create a CoalesceRDD. Instead it will reduce the number of partitions by adjusting the partitioning of C* data on read. Using this method will override spark.cassandra.input.split.size."

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala [hunk 7]
```diff
@@ -221,10 +238,36 @@ class CassandraTableScanRDD[R] private[connector](
...
+    *
+    * @param numPartitions number of partinions
+    * @param shuffle       weither to call shufle after
+    * @param ord
+    * @return new CassandraTableScanRDD with predefined nube of paritions
+    */
+
+  override def coalesce(numPartitions: Int, shuffle: Boolean = false)(implicit ord: Ordering[R] = null)
+  : RDD[R] = {
+    val rdd = copy(numPartitions = Some(numPartitions))
```
Comment (RussellSpitzer) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableScanRDD.scala]: I think this should be a `copy(readConf = readConf.copy(splitCount = Some(numPartitions))`

**Comment by artem-aliev** [comment]
@RussellSpitzer I have addressed your comments: removing code is the only way to reduce number of bugs :)

**Comment by RussellSpitzer** [comment]
+1

**Comment by bcantoni** [comment]
Triage: Artem will rebase off latest master (for 2.0).

Also please add user docs if that makes sense.
